### PR TITLE
RS API Extension - Network Energy Core

### PR DESCRIPTION
**Pull request by samtrion**
+ added also Config option to adjust max receive rate for the controller

### CODE REVIEW

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
Thank you for your PR, but there is still some work to be done.

**Comment by Samtrion** [comment]
Thanks for the review, i will fix this 😀

File: src/main/java/com/raoulvdberge/refinedstorage/RSConfig.java [hunk 1]
```diff
@@ -140,6 +141,7 @@ public void onConfigChangedEvent(ConfigChangedEvent.OnConfigChangedEvent event)
     private void loadConfig() {
         //region Energy
         controllerBaseUsage = config.getInt("controllerBase", ENERGY, 0, 0, Integer.MAX_VALUE, "The base energy used by the Controller");
+        controllerMaxReceive = config.getInt("controllerMaxReceive", ENERGY, Integer.MAX_VALUE, 0, Integer.MAX_VALUE, "The maximum RF the controller receives per Tick");
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/RSConfig.java]: Change RF into "energy" and lowercase "Tick"

File: src/main/java/com/raoulvdberge/refinedstorage/api/network/INetwork.java [hunk 2]
```diff
@@ -26,7 +28,7 @@
     /**
      * @return the energy usage per tick of this network
      */
-    int getEnergyUsage();
+    int getNetworkEnergyUsage();
```
[the remark this answers is not in the document]
Comment (Samtrion) [reply src/main/java/com/raoulvdberge/refinedstorage/api/network/INetwork.java]: To differentiate between `INetwork.getEnergyUsage()` and `INetworkNode.getEnergyUsage()` to exclude possible side effects.

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergyCore.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.raoulvdberge.refinedstorage.api.energy;
+
+import java.util.UUID;
+
+import net.minecraftforge.energy.IEnergyStorage;
+
+public interface IEnergyCore {
```
[the remark this answers is not in the document]
Comment (Samtrion) [reply src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergyCore.java]: Surely the name `IEnergy` is enough.
As part of this, I would also rename 'EnergyForgeCoreProxy' to 'EnergyProxy' and 'EnergyForgeCore' to 'Energy'

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergyCore.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.raoulvdberge.refinedstorage.api.energy;
+
+import java.util.UUID;
+
+import net.minecraftforge.energy.IEnergyStorage;
```
[the remark this answers is not in the document]
Comment (Samtrion) [reply src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergyCore.java]: That's a leftover of a test. :(

File: src/main/java/com/raoulvdberge/refinedstorage/api/network/INetwork.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Samtrion) [reply src/main/java/com/raoulvdberge/refinedstorage/api/network/INetwork.java]: But that is obsolete as the nodes control the energy consum.

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/EnergyProxy.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package com.raoulvdberge.refinedstorage.api.energy;
+
+import net.minecraftforge.energy.IEnergyStorage;
+
+public final class EnergyProxy implements IEnergyStorage {
+	private final int maxReceive;
+	private final int maxExtract;
+	private final IEnergy energy;
+
+	public EnergyProxy(IEnergy energyCore, int maxTransfer) {
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/energy/EnergyProxy.java]: Rename to energy

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/EnergyProxy.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+		this(energyCore, maxTransfer, maxTransfer);
+	}
+
+	public EnergyProxy(IEnergy energyCore, int maxReceive, int maxExtract) {
+		this.energy = energyCore;
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/energy/EnergyProxy.java]: Rename energyCore to energy

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.raoulvdberge.refinedstorage.api.energy;
+
+import java.util.UUID;
+
+public interface IEnergy {
+	void decreaseCapacity(UUID id, int amount);
+
+	int extract(int amount, boolean simulate);
+
+	int getMaxEnergy();
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java]: getCapacity

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+	int getStored();
+
+	void increaseCapacity(UUID id, int amount);
+
+	int receive(int amount, boolean simulate);
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java]: insert (to be analog with "extract")

File: src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+	void setStored(int energyAmount);
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/api/energy/IEnergy.java]: energy instead of energyAmount

**Comment by Samtrion** [comment]
Sorry, I'm still fighting a bit with the JAVA IDE of my choice.
Hope I have now treated all Stylebreaks.

**Pull request merged**